src/metaxy/metadata_store/base.py

```diff
@@ -494,30 +494,57 @@ def write_metadata(
         if not is_system_table:
             self._validate_project_write(feature)
 
-        # Convert Narwhals to Polars if needed
-        if isinstance(df, nw.DataFrame):
-            df = df.to_polars()
-        # nw.DataFrame also matches as DataFrame in some contexts, ensure it's Polars
-        if not isinstance(df, pl.DataFrame):
-            # Must be some other type - shouldn't happen but handle defensively
-            if hasattr(df, "to_polars"):
-                df = df.to_polars()
-            elif hasattr(df, "to_pandas"):
-                df = pl.from_pandas(df.to_pandas())
-            else:
-                raise TypeError(f"Cannot convert {type(df)} to Polars DataFrame")
+        # Convert Narwhals to native Polars (both lazy and eager)
+        if isinstance(df, (nw.LazyFrame, nw.DataFrame)):
+            df = df.to_native()  # type: ignore[assignment]
```

Comment:
This won't convert to polars if the frame is backed by let's say Ibis. 

You have to `.lazy().collect().to_polars()`. We already have a helper for that in _utils.py if I recall correctly. 

It's also best if we issue a user warning in case we need to collect a DataFrame like this (should only happen for lazy frames not backed by Polars). 

---

I almost got this logic into main anyway, so perhaps don't bother right now, it will be fixed anyway.